stb_image fuzzing fixes

rcgordon opened the pull request:
This PR has all the hardening work I did on stb_image:

- Fuzzing inputs with AFL, using a 2x2 image in bmp, gif, jpg, hdr, png, pnm, psd, and tga. (pic not tested). Those files are available in [stbi_afl.zip](https://github.com/nothings/stb/files/4547882/stbi_afl.zip) .
- Running this testsuite through Valgrind and AddressSanitizer: https://code.google.com/archive/p/imagetestsuite/downloads

Testing was done with and without SSE2 support.

randy408 commented:
There's a PNG-only fuzzer on [OSS-Fuzz](https://github.com/google/oss-fuzz) that found [this](https://github.com/nothings/stb/pull/918) which doesn't seem to be addressed by this PR. I'm the only one who looks at the bug reports for that fuzzer so It hasn't been expanded to other formats because I reckoned there would too many  bug reports for me to handle.

If you're willing to address the bug reports that would come up from increasing code coverage I could add you as a contact for the project. Overall it's a simple process, issues are automatically created on chromium.org with test cases and are auto-closed when the fix is in master.

rcgordon commented:
> I maintain a PNG-only fuzzer on [OSS-Fuzz](https://github.com/google/oss-fuzz) that found [this](https://github.com/nothings/stb/pull/918) which doesn't seem to be addressed by this PR.

I added your fix to this PR (commit credited to you, of course!). It's possible other changes made mitigated the immediate need for this fix, and it's also possible my fuzzing just missed it, but it looked like a correct, localized and reasonable change in any case, so I added it in.

> If you're willing to address the bug reports that would come up from increasing code coverage I could add you as a contact for the project.

Thanks, but I don't have time to take on that responsibility.

bjornbytes commented on stb_image.h:
```diff
@@ -4079,7 +4112,12 @@ static int stbi__zhuffman_decode_slowpath(stbi__zbuf *a, stbi__zhuffman *z)
-   if (a->num_bits < 16) stbi__fill_bits(a);
+   if (a->num_bits < 16) {
+      if (stbi__zeof(a)) {
+         return -1;   /* report error for unexpected end of data. */
```
Is it possible that this is too aggressive now?  I'm using stb_image to decompress files in zip archives (long story) and I started seeing some unexpected failures after this change.  When this error condition is hit, the zbuf has 13 bits in it and it's trying to decode the last byte of the stream.  Still brushing up on huffman stuff so I might be missing something obvious.  I can scrape together a test case if that's helpful

nothings replied:
What if you add an stbi_fill_bits(a) after the if num_bits and before the if stbi_zeof?

nothings replied:
actually I think this is going to need a messier fix, since it can potentially intentionally "read past the end of file" while trying to do the "fast" huffman table lookup. ugh, need to figure out the right way to do it.
